### PR TITLE
ORDER BY Arrays and Objects

### PR DESCRIPTION
**Pull request by bchong95**
# ORDER BY Arrays and Objects

## Description

Adds support to ORDER BY Arrays and Objects. The algorithm is arrays come before objects but after strings. comparing two arrays or two objects is done by comparing their hashes.

Internal PR:
https://msdata.visualstudio.com/_git/CosmosDB/commit/952ea1a412850770e417b892c25a11272fd119cb?path=%2FProduct%2FSDK%2F.net%2FMicrosoft.Azure.Documents.Client%2FQuery%2FDistinctMap.UnorderedDistinctMap.cs&gridItemType=2&mpath=%2FProduct%2FSDK%2F.net%2FMicrosoft.Azure.Documents.Client%2FQuery%2FDistinctMap.UnorderedDistinctMap.cs&opath=%2FProduct%2FSDK%2F.net%2FMicrosoft.Azure.Documents.Client%2FQuery%2FDistinctMap.UnorderedDistinctMap.cs&mversion=GC952ea1a412850770e417b892c25a11272fd119cb&oversion=GCec11d7d0b0246a5f3d53536f0e10d0d2ed50d0e5&_a=compare

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Query/Core/DistinctHash.cs [hunk 1]
```diff
@@ -0,0 +1,316 @@
+﻿// ------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.  All rights reserved.
+// ------------------------------------------------------------
+
+namespace Microsoft.Azure.Cosmos.Query
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Query/Core/DistinctHash.cs]: Should this be added to the changelog? I don't see any feature switches blocking it.

**Comment by kirankumarkolli** [comment]
Cosmos is schema less -> in query a single "ORDER BY <columnname>" might result with different data-types for <columnname>, hence the type ordering. Is my understanding right?

Is the order defined unique to Cosmos or some standard defined?

File: Microsoft.Azure.Cosmos/src/Query/Core/ItemComparer.cs [hunk 2]
```diff
@@ -207,25 +215,25 @@ private static int TypeToOrder(CosmosElement cosmosElement)
                                 order = 2;
                                 break;
                             case CosmosNumberType.Float32:
-                                order = 6;
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Query/Core/ItemComparer.cs]: Just to re-assert, the order change impact is localized to process only right? i.e. any possible back-compact issues?

File: Microsoft.Azure.Cosmos/src/Query/Core/DistinctHash.cs [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+    using System.Text;
+    using Microsoft.Azure.Cosmos.CosmosElements;
+
+    /// <summary>
+    /// Base class for DistinctHash.
+    /// This class is able to take hashes with seeded values.
+    /// </summary>
+    internal static class DistinctHash
+    {
+        private static readonly UInt192 RootHashSeed = UInt192.Create(
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Query/Core/DistinctHash.cs]: Are these seeds BE defined?

File: Microsoft.Azure.Cosmos/src/Query/Core/DistinctHash.cs [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+            ulong hash64 = Microsoft.Azure.Documents.Routing.MurmurHash3.Hash64(bytes, bytes.Length, seed.GetHigh());
+            return UInt192.Create(hash128.GetLow(), hash128.GetHigh(), hash64);
+        }
+
+        /// <summary>
+        /// Gets the hash of a JToken value.
+        /// </summary>
+        /// <param name="cosmosElement">The element to load.</param>
+        /// <returns>The hash of the JToken.</returns>
+        public static UInt192 GetHash(CosmosElement cosmosElement)
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Query/Core/DistinctHash.cs]: Why others needs to be public? ex: GetHash(byte[] bytes, UInt192 seed)

File: Microsoft.Azure.Cosmos/src/Query/Core/ItemComparer.cs (hunk 2 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Query/Core/ItemComparer.cs]: Order within type is defined. We haven't defined order across types. any back-compat issues are due to relying on undefined behavior.

File: Microsoft.Azure.Cosmos/src/Query/Core/DistinctHash.cs (hunk 3 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Query/Core/DistinctHash.cs]: Yeah

File: Microsoft.Azure.Cosmos/src/Query/Core/DistinctHash.cs (hunk 4 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Query/Core/DistinctHash.cs]: Not sure. Probably someone calls. Ill double check.

File: Microsoft.Azure.Cosmos/src/Query/Core/DistinctHash.cs (hunk 1 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Query/Core/DistinctHash.cs]: Ill add it to the changelog.

File: Microsoft.Azure.Cosmos/src/Query/Core/DistinctHash.cs [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+    /// This class is able to take hashes with seeded values.
+    /// </summary>
+    internal static class DistinctHash
+    {
+        private static readonly UInt192 RootHashSeed = UInt192.Create(
+            0xddef2a91418ae4ea,
+            0x9fcd55808a038bf3,
+            0x3e2614f44255738b);
+
+        private static readonly UInt192 NullHashSeed = UInt192.Create(
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Query/Core/DistinctHash.cs]: Please fix the test failures.

File: Microsoft.Azure.Cosmos/src/Query/Core/DistinctHash.cs (hunk 4 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Query/Core/DistinctHash.cs]: Yeah someone calls it.

File: templates/emulator-setup.yml [hunk 6]
```diff
@@ -11,7 +11,7 @@ steps:
         mkdir "$env:temp\Azure Cosmos DB Emulator"
         lessmsi x "$env:temp\azure-cosmosdb-emulator.msi" "$env:temp\Azure Cosmos DB Emulator\"
         Write-Host "Starting Comsos DB Emulator" -ForegroundColor green 
-        Start-Process "$env:temp\Azure Cosmos DB Emulator\SourceDir\Azure Cosmos DB Emulator\CosmosDB.Emulator.exe" "/NoExplorer /NoUI /DisableRateLimiting /PartitionCount=100 /Consistency=Strong /enableRio /overrides=sqlAllowGroupByClause:true" -Verb RunAs
+        Start-Process "$env:temp\Azure Cosmos DB Emulator\SourceDir\Azure Cosmos DB Emulator\CosmosDB.Emulator.exe" "/NoExplorer /NoUI /DisableRateLimiting /PartitionCount=100 /Consistency=Strong /enableRio /overrides=sqlAllowGroupByClause:true;enableDefaultIndexEncodingOptions:true" -Verb RunAs
```
Comment (j82w) [inline templates/emulator-setup.yml]: Will this change require a new emulator version to be published for the feature to work?

**Comment by kirankumarkolli** [comment]
Please create feature issues on non .NET SDK repos' for tracking.

File: templates/emulator-setup.yml (hunk 6 above)
Comment (bchong95) [reply templates/emulator-setup.yml]: I think the existing public emulator already has this flag.

Comment (j82w) [reply templates/emulator-setup.yml]: If this is a public feature we can't expect customer to use a undocumented flag.

File: changelog.md [hunk 7]
```diff
@@ -12,6 +12,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - [#814](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/814) Ability to limit to configured endpoint only
 - [#822](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/822) GROUP BY query support.
 - [#844](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/844) Added PartitionKeyDefinitionVersion to container builder
+- [#853](https://github.com/Azure/azure-cosmos-dotnet-v3/pull/853) ORDER BY Arrays and Object support.
```
Comment (j82w) [inline changelog.md]: Please create a new Unreleased section and add the changelog there.

**Pull request merged**